File: sdk/php/src/Webhooks.php
<?php
declare(strict_types=1);

namespace MerchantPlatform;

interface WebhookInbox { public function process(string $eventId, string $bodyDigest, callable $handler): string; }
final class VerifiedWebhook { public function __construct(public readonly WebhookEvent $event, public readonly string $eventId, public readonly string $keyId, public readonly int $timestamp, public readonly string $bodyDigest) {} }
final class WebhookVerificationException extends \RuntimeException {}
final class Webhooks {
    public static function verify(string $rawBody, string $signatureHeader, string $contentDigest, callable $resolveSecret, ?int $now = null, int $toleranceSeconds = 300): VerifiedWebhook {
        $parts = []; foreach (explode(',', $signatureHeader) as $part) { [$key,$value] = array_pad(explode('=', trim($part), 2), 2, null); if ($value !== null) $parts[$key] = $value; }
        if (!isset($parts['t'],$parts['key'],$parts['event'],$parts['v1']) || !ctype_digit($parts['t'])) throw new WebhookVerificationException('invalid webhook signature header');
        $timestamp = (int)$parts['t']; if (abs(($now ?? time()) - $timestamp) > $toleranceSeconds) throw new WebhookVerificationException('webhook timestamp outside tolerance');
        $digest = 'sha-256=:'.base64_encode(hash('sha256', $rawBody, true)).':'; if (!hash_equals($digest, $contentDigest)) throw new WebhookVerificationException('webhook content digest mismatch');
        $secret = $resolveSecret($parts['key']); if (!is_string($secret) || $secret === '') throw new WebhookVerificationException('unknown webhook key');
        $expected = Signing::base64Url(hash_hmac('sha256', $parts['event'].'.'.$timestamp.'.'.$rawBody, $secret, true)); if (!hash_equals($expected, $parts['v1'])) throw new WebhookVerificationException('webhook signature mismatch');
        try { $data = json_decode($rawBody, true, 512, JSON_THROW_ON_ERROR); } catch (\JsonException) { throw new WebhookVerificationException('invalid webhook JSON'); }
        if (($data['event_id'] ?? null) !== $parts['event'] || ($data['schema_version'] ?? null) !== '1') throw new WebhookVerificationException('webhook envelope mismatch');
        return new VerifiedWebhook(WebhookEvent::fromArray($data), $parts['event'], $parts['key'], $timestamp, $digest);
    }
    public static function acknowledgement(string $eventId): array { return ['acknowledged_event_id'=>$eventId]; }
}

File: sdk/dotnet/src/MerchantPlatform/Webhooks.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace MerchantPlatform;

public enum InboxResult { Processed, Duplicate, Conflict }
public interface IWebhookInbox<TTransaction> { Task<InboxResult> ProcessAsync(string eventId,string bodyDigest,Func<TTransaction,Task> handler,CancellationToken cancellationToken=default); }
public sealed record VerifiedWebhook(WebhookEvent Event,string EventId,string KeyId,long Timestamp,string BodyDigest);
public sealed class WebhookVerificationException(string message):Exception(message);
public static class Webhooks {
    private static readonly JsonSerializerOptions Json=new(){PropertyNamingPolicy=JsonNamingPolicy.SnakeCaseLower,PropertyNameCaseInsensitive=true};
    public static VerifiedWebhook Verify(ReadOnlySpan<byte> rawBody,string signatureHeader,string contentDigest,Func<string,string?> resolveSecret,DateTimeOffset? now=null,long toleranceSeconds=300){var parts=new Dictionary<string,string>();foreach(var item in signatureHeader.Split(',')){var pair=item.Trim().Split('=',2);if(pair.Length==2)parts[pair[0]]=pair[1];}if(!parts.TryGetValue("t",out var stamp)||!long.TryParse(stamp,out var timestamp)||!parts.TryGetValue("key",out var keyId)||!parts.TryGetValue("event",out var eventId)||!parts.TryGetValue("v1",out var provided))throw new WebhookVerificationException("invalid webhook signature header");if(Math.Abs((now??DateTimeOffset.UtcNow).ToUnixTimeSeconds()-timestamp)>toleranceSeconds)throw new WebhookVerificationException("webhook timestamp outside tolerance");var digest=Signing.Digest(rawBody);if(!Fixed(digest,contentDigest))throw new WebhookVerificationException("webhook content digest mismatch");var secret=resolveSecret(keyId);if(string.IsNullOrEmpty(secret))throw new WebhookVerificationException("unknown webhook key");using var hmac=new HMACSHA256(Encoding.UTF8.GetBytes(secret));var prefix=Encoding.UTF8.GetBytes($"{eventId}.{timestamp}.");var input=new byte[prefix.Length+rawBody.Length];prefix.CopyTo(input,0);rawBody.CopyTo(input.AsSpan(prefix.Length));var expected=Signing.Base64Url(hmac.ComputeHash(input));if(!Fixed(expected,provided))throw new WebhookVerificationException("webhook signature mismatch");WebhookEvent? value;try{value=JsonSerializer.Deserialize<WebhookEvent>(rawBody,Json);}catch(JsonException){throw new WebhookVerificationException("invalid webhook JSON");}if(value is null||value.EventId!=eventId||value.SchemaVersion!="1")throw new WebhookVerificationException("webhook envelope mismatch");return new VerifiedWebhook(value,eventId,keyId,timestamp,digest);}
    public static IReadOnlyDictionary<string,string> Acknowledgement(string eventId)=>new Dictionary<string,string>{{"acknowledged_event_id",eventId}};
    private static bool Fixed(string left,string right){var a=Encoding.UTF8.GetBytes(left);var b=Encoding.UTF8.GetBytes(right);return a.Length==b.Length&&CryptographicOperations.FixedTimeEquals(a,b);}
}
